File: plugins/collector/event/apps/bigdata/zookeeper.go
package bigdata

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"regexp"
	"strings"
)

type Zookeeper struct {
	version    string
	reg        *regexp.Regexp
	regVersion *regexp.Regexp
}

func (Zookeeper) Name() string { return "zookeeper" }

func (Zookeeper) Type() string { return "bigdata" }

func (z Zookeeper) Version() string { return z.version }

func (z *Zookeeper) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	if z.reg == nil {
		z.reg = regexp.MustCompile(` org\.apache\.zookeeper\.`)
	}
	if res := z.reg.FindString(p.Argv); res == "" {
		return false
	}
	return true
}

func (z *Zookeeper) Run(p *process.Process) (m map[string]string, err error) {
	if z.regVersion == nil {
		z.regVersion = regexp.MustCompile(`zookeeper-(\d+\.)+(\d+)\.jar`)
	}
	// get the version
	var fds []string
	if fds, err = p.Fds(); err != nil {
		return
	}
	for _, fd := range fds {
		if filepath.Ext(fd) != ".jar" {
			continue
		}
		if res := z.regVersion.FindString(filepath.Base(fd)); res != "" {
			z.version = strings.TrimRight(strings.TrimPrefix(res, "zookeeper-"), ".jar")
			break
		}
	}
	return
}

func init() { apps.Regist(&Zookeeper{}) }

File: plugins/collector/src/event/apps/bigdata/zookeeper.rs
use crate::event::apps::{find_version, jar_names_for_pid, AppProc, IApp};
use std::collections::HashMap;

pub struct Zookeeper {
    version: String,
}

impl Zookeeper {
    pub fn new() -> Self { Zookeeper { version: String::new() } }
}

impl IApp for Zookeeper {
    fn name(&self)     -> &'static str { "zookeeper" }
    fn app_type(&self) -> &'static str { "bigdata" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java" && p.argv.contains("org.apache.zookeeper.")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        // "zookeeper-3.8.0.jar"
        for jar in jar_names_for_pid(p.pid) {
            if jar.starts_with("zookeeper-") && jar.ends_with(".jar") {
                let mid = jar.trim_start_matches("zookeeper-").trim_end_matches(".jar");
                let ver = find_version(mid);
                if !ver.is_empty() { self.version = ver; break; }
            }
        }
        Ok(HashMap::new())
    }
}
